### PR TITLE
Implement DIP1029: Add `throw` as Function Attribute

### PR DESCRIPTION
**Pull request by ljmf00**
https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1029.md

Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

- [x] Specification PR (https://github.com/dlang/dlang.org/pull/3082)
- [x] DMD Implementation
- [x] Tests
  - [x] Header generation of `throw`
  - [x] Templated functions: do they infer default or `throw`? how does that interact with toplevel `throw`/`nothrow`?
  - [ ] throw virtual functions that are overriden with `nothrow`
  - [x] more cases for `throw:`/`nothrow:` interacting with function/template declarations
  - [x] Add tests with `__traits(getFunctionAttributes, func)`
  - [ ] Cast `nothrow` to `throw`
  - [ ] Check for forbidden cast of `throw` to `nothrow`

CC: @WalterBright

### CODE REVIEW

**Comment by ljmf00** [comment]
Buildkite seems failing due to some side effects, not related to this PR. Although, need to write some more tests to cover all this DIP. For now, I would appreciate a review of my implementation approach.

File: test/compilable/dip1029.d [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    override void foo() @nogc;
+}
+
+static assert(is(typeof(&CA.foo) == void function() throw));
+static assert(is(typeof(&CB.foo) == void function() throw @nogc));
+
+
+// inferred types and attributes for lambdas
+static assert(is(typeof(() {}) == void function() pure nothrow @nogc @safe));
+static assert(is(typeof(() { throw new Exception(""); }) == void function() pure @safe));
```
Comment (RazvanN7) [inline test/compilable/dip1029.d:41]: why is this not inferred as `throw` wheres the subsequent one is?

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
This still requires some tests for:

- header generation of throw
- templated functions: do they infer default or throw? how does that interact with toplevel throw/nothrow?
- throw virtual functions that are overriden with nothrow
- more cases for `throw:`/`nothrow:` interacting with function/template declarations

The DIP does not specify whether inference should deduce throw or default for functions that are not nothrow. I would argue that a function has to be throw or nothrow; default is just an intermediary state where the compiler does not know but should assume throw. As a precedent, in the case of safety, `@system` is specifically deduced when inference occurs. In this implementation there seems to be a gap between throw and default. Looking at `test/compilable/dip1029.d` line 41 it seems that inference does not deduce throw, but default. This is important for things like header generation, error messages, typeof etc.

File: test/compilable/dip1029.d (hunk 1 above)
Comment (ljmf00) [reply test/compilable/dip1029.d:41]: This is inferred as `throw`, `void function() pure throw @safe` is the same as `void function() pure @safe`.

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -4981,18 +4985,22 @@ extern (C++) final class TypeFunction : TypeNext
...
 
-    /// set or get if the function has the `nothrow` attribute
-    bool isnothrow() const pure nothrow @safe @nogc
-    {
-        return (funcFlags & FunctionFlag.isnothrow) != 0;
-    }
-    /// ditto
-    void isnothrow(bool v) pure nothrow @safe @nogc
+
+    deprecated("This property was removed in favor of throw_ field")
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/mtype.d:4989]: I thought that in case someone uses dmd as a library. I hear that was a thing, so I didn't remove it all to prevent breaking user code, although I don't know its state.

File: src/dmd/mtype.d [hunk 3]
```diff
@@ -4981,18 +4985,22 @@ extern (C++) final class TypeFunction : TypeNext
...
     {
-        if (v) funcFlags |= FunctionFlag.isnothrow;
-        else funcFlags &= ~FunctionFlag.isnothrow;
+        /// set or get if the function has the `nothrow` attribute
+        bool isnothrow() const pure nothrow @safe @nogc
+        {
+            return throw_ == THROW.nothrow_ || throw_ == THROW.default_;
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/mtype.d:4994]: You are right. Need to add a test case for this, if `isnothrow` is kept. The compiler itself no longer uses `isnothrow` function. It is only to give backwards compatibility and prevent breaking code, as I described above.

File: src/dmd/mtype.d [hunk 4]
```diff
@@ -4981,18 +4985,22 @@ extern (C++) final class TypeFunction : TypeNext
...
+        }
+        /// ditto
+        void isnothrow(bool v) pure nothrow @safe @nogc
+        {
+            throw_ = v ? THROW.nothrow_ : THROW.throw_;
+        }
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/mtype.d:5000]: The compiler itself is no longer using it.

File: src/dmd/parse.d [hunk 5]
```diff
@@ -4355,6 +4369,10 @@ class Parser(AST) : Lexer
                 stc = STC.nothrow_;
                 goto L1;
 
+            case TOK.throw_:
+                stc = STC.throw_;
+                goto L1;
+
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/parse.d:4375]: I still need to check which section this code belongs. I arranged the code based on existing attributes. I will try to write some more tests to cover all these warnings.

**Comment by ljmf00** [comment]
> This still requires some tests for:
> 
> * header generation of throw
> * templated functions: do they infer default or throw? how does that interact with toplevel throw/nothrow?
> * throw virtual functions that are overriden with nothrow
> * more cases for `throw:`/`nothrow:` interacting with function/template declarations

Thanks for the list :+1: . I'm going to do that.

> The DIP does not specify whether inference should deduce throw or default for functions that are not nothrow. I would argue that a function has to be throw or nothrow; default is just an intermediary state where the compiler does not know but should assume throw. As a precedent, in the case of safety, `@system` is specifically deduced when inference occurs. In this implementation there seems to be a gap between throw and default. Looking at `test/compilable/dip1029.d` line 41 it seems that inference does not deduce throw, but default. This is important for things like header generation, error messages, typeof etc.

I choose to use a default mechanism to prevent breaking changes.

I thought that this was not specified in the DIP because it is implicitly implied by "no breaking changes". It is important to test this behaviour because existing code might get the `typeof` a `throw` function without specifying any attribute and adding `throw` attribute to the returned type may introduce a breaking change.

```d
void foo() {}
void bar() @system {}
```

Doing a `typeof` of `foo` and `bar` produces different output and if the user compares it with, for example, a .stringof, this may break it.

I also see this as acceptable because this is how the `@system` attribute comparison works with the default value.

```d
is(void function() == void function() @system) // true
```

AFAIK, the intermediate state where the compiler can't assume throw is validated by `FUNCFLAG.nothrowInprocess` rather than `THROW.default_`.

Tell me if I interpreted your comment/suggestion wrongly.

File: src/dmd/mtype.d (hunk 2 above)
[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/mtype.d:4989]: Technically, it could be used, but I doubt that. I guess there's no harm in using deprecatins here.

**Comment by RazvanN7** [comment]
> Doing a typeof of foo and bar produces different output and if the user compares it with, for example, a .stringof, this may break it.

In your example there is no inference. In that case, I agree that typeof should return whatever the explicit annotation is. However, in the case of delegates/lambdas it's not that obvious: these always have a body and their attributes are explicitly embedded in the type (the same as `@system`).

Also, for template functions, I think that `throw` should be explicitly added as it is the case for `system` however that might break some code that uses .stringof the way you mentioned.

To summarize: should function declarations that have their attributes inferred (deletgates, lambdas, function templates) explicitly be marked as `throw` or be left unmarked? On the long run, the compiler should explicitly annotate with `throw`, however, this might break some code (when generating headers that contain delegates/lambdas, when checking the type of template instances via .stringof etc.)

cc @WalterBright

**Comment by ljmf00** [comment]
> > Doing a typeof of foo and bar produces different output and if the user compares it with, for example, a .stringof, this may break it.
> 
> In your example there is no inference. In that case, I agree that typeof should return whatever the explicit annotation is. However, in the case of delegates/lambdas it's not that obvious: these always have a body and their attributes are explicitly embedded in the type (the same as `@system`).
> 
> Also, for template functions, I think that `throw` should be explicitly added as it is the case for `system` however that might break some code that uses .stringof the way you mentioned.
> 
> To summarize: should function declarations that have their attributes inferred (deletgates, lambdas, function templates) explicitly be marked as `throw` or be left unmarked? On the long run, the compiler should explicitly annotate with `throw`, however, this might break some code (when generating headers that contain delegates/lambdas, when checking the type of template instances via .stringof etc.)

I agree. We could add a `-preview=dip1029` to express that behaviour. I'm not aware of the preview process, but I assume is the same as the deprecation process but inverted, right?

**Comment by ljmf00** [comment]
This code will still have breaking changes. `__traits(getFunctionAttributes)` gives you `@system` on a function without parameters.

**Comment by pbackus** [comment]
> however that might break some code that uses .stringof the way you mentioned.

As far as I know the language spec makes no guarantees whatsoever about the specific contents of `.stringof`, so anyone who relies on it in this way is doing so at their own risk.

**Comment by ljmf00** [comment]
Needs #13121 for trailing whitespace free tests.

**Comment by ljmf00** [comment]
> #13121 has been merged.

Yes, but I still need to commit the header generation tests and fix the logic for it.

File: src/dmd/dmangle.d [hunk 6]
```diff
@@ -510,7 +510,7 @@ public:
 
         if (ta.purity)
             buf.writestring("Na");
-        if (ta.isnothrow)
+        if (ta.throw_ == THROW.nothrow_)
```
Comment (kubo39) [inline src/dmd/dmangle.d:513]: I think you need to add a mangling rule for `THROW.throw_` and tests for `__traits(getFunctionAttributes, ...)`.

File: src/dmd/mtype.d [hunk 7]
```diff
@@ -4149,6 +4147,12 @@ extern (C++) final class TypeFunction : TypeNext
         if (stc & STC.scopeinferred)
             this.isscopeinferred = true;
 
+        this.throw_ = THROW.default_;
+        if (stc & STC.nothrow_)
+            this.throw_ = THROW.nothrow_;
+        if (stc & STC.throw_)
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/mtype.d]: Done

File: src/dmd/mtype.d [hunk 8]
```diff
@@ -4149,6 +4147,12 @@ extern (C++) final class TypeFunction : TypeNext
         if (stc & STC.scopeinferred)
             this.isscopeinferred = true;
 
+        this.throw_ = THROW.default_;
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/mtype.d]: Done. Going to do that for trust too.

File: src/dmd/dsymbolsem.d [hunk 9]
```diff
@@ -2976,8 +2976,12 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                 sc.stc |= STC.ref_;
             if (tf.isScopeQual)
                 sc.stc |= STC.scope_;
-            if (tf.isnothrow)
+            if (tf.throw_ != THROW.default_)
+                sc.stc &= ~STC.throwGroup;
+            if (tf.throw_ == THROW.nothrow_)
```
[the remark this answers is not in the document]
Comment (ljmf00) [reply src/dmd/dsymbolsem.d]: Done.

File: src/dmd/dmangle.d (hunk 6 above)
Comment (ljmf00) [reply src/dmd/dmangle.d:513]: What do you mean? There is no need to add new mangling for `throw`. The same as trust, only `safe` and `trusted` has mangling rules, `system` is considered the default, and therefore no mangling rule is needed.

**Comment by ljmf00** [comment]
I added the rest of the test cases. I just need to add cast() and virtual functions.

> This code will still have breaking changes. `__traits(getFunctionAttributes)` gives you `@system` on a function without ~parameters~ attributes.

As I suspected, this fails to compile on Phobos. Should I add a flag to the compiler to make the transition or make `__traits(getFunctionAttributes)` hide `throw` on `THROW.default_`? That will make it a special case tho. How can we proceed here @RazvanN7 ?

File: test/compilable/dip1029.d (hunk 1 above)
Comment (ljmf00) [reply test/compilable/dip1029.d:41]: I added a check on line 45 to prove it.

**Comment by RazvanN7** [comment]
We have the following solutions:

- omit `throw` when it is not specifically added by the user. This would be the easiest way, however, it will be inconsistent with what happens in the case of `@system/@safe`.
- issue a deprecation whenever `getFunctionAttributes` is used on a `throw` function that is not annotated as such.
- fix the breakages and go with it as is.

Either way we choose it's a bit nasty.  I, personally, would be inclined to go the latter route since I expect that the way phobos uses `getFunctionAttributes` is not a common occurrence. But, I think that there are folks who might not necessarily agree with me, so issuing a deprecation is probably the best way to go here.

**Comment by RazvanN7** [comment]
Another alternative would be to deprecate `getFunctionAttributes` altogether for a newer `getFuncAttrs` which outputs `throw`, however, this is not scalable if we intend to implement `impure` and `gc`

**Comment by ibuclaw** [comment]
1. Would this be a precursor for diagnostics akin to: `error: func() throws but is not caught`
2. Possibility of future extensions to allow specifying which exception is thrown, e.g: `int openFile() throw(FileException) { ... }`

**Comment by ljmf00** [comment]
> 1. Would this be a precursor for diagnostics akin to: `error: func() throws but is not caught`

I don't think so, the default, which is the "hidden" `throw` doesn't behave like that, why would with specified `throw`? I would argue that makes sense if we specified the Exception class, explicitly, but this DIP doesn't cover that.

> 2. Possibility of future extensions to allow specifying which exception is thrown, e.g: `int openFile() throw(FileException) { ... }`

This DIP only creates a contrary attribute to negate `nothrow`.

**Comment by ljmf00** [comment]
> fix the breakages and go with it as is.

I can create an attempt to fix the breakage on Phobos, but that doesn't guarantee the fact that this will break user code, although the user should be prepared for those changes, as any other attribute might get introduced. To do this attempt, should I add a synchronization/transition flag to do that transition? If yes, what flag? My suggestion would be to stick with a `-revert=` flag, that way users can revert that behaviour to the old one, preventing breakage and in the long term, deprecate that flag when we have `impure` and `@gc`?

**Comment by ljmf00** [comment]
Ping @RazvanN7 :)

**Comment by RazvanN7** [comment]
The more I think about this, the more I feel that a deprecation is a more suitable alternative. We added a new function attribute, people need to update their code to take it into account.

**Comment by ljmf00** [comment]
> The more I think about this, the more I feel that a deprecation is a more suitable alternative. We added a new function attribute, people need to update their code to take it into account.

I can agree with the deprecation if we implement this to the other attributes, I would say.

**Comment by RazvanN7** [comment]
Yes, definitely, if other attributes will be implemented then the same deprecation path must be taken.

**Comment by ljmf00** [comment]
> Yes, definitely, if other attributes will be implemented then the same deprecation path must be taken.

Ah, so you mean this?

> - issue a deprecation whenever getFunctionAttributes is used on a throw function that is not annotated as such by the user.

I thought you were talking about implementing the new trait and deprecate the older one. But yeah, I share the same opinion, if that's what you mean.

**Comment by ljmf00** [comment]
> > Ah, so you mean this?
> 
> Yes.

Ok, I can go with implementing it that way. I think this is the best way among the other options we have.

**Comment by RazvanN7** [comment]
ping @ljmf00

**Comment by RazvanN7** [comment]
@ljmf00 any updates on this? This is so close to completion, it's a pity it got stalled so much.
